### PR TITLE
Fix a hang in `tests/store/tracking/test_sqlalchemy_store.py`

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/12560?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12560/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12560
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

A temporary solution for the following error:

https://github.com/mlflow/mlflow/actions/runs/9775455006/job/26985875566

```
+ [[ ***mysql:3306/mlflowdb?charset=utf8mb4 == mssql* ]]
+ exec pytest tests/store/tracking/test_sqlalchemy_store.py tests/store/model_registry/test_sqlalchemy_store.py tests/db
Environment variable MLFLOW_TRACKING_URI is set to '***mysql:3306/mlflowdb?charset=utf8mb4', which may interfere with tests.
============================= test session starts ==============================
platform linux -- Python 3.8.19, pytest-8.2.2, pluggy-1.5.0 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /mlflow/home
configfile: pytest.ini
plugins: cov-5.0.0
collecting ... collected 273 items
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[sqlite-pysqlite] PASSED [  0%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[sqlite-pysqlcipher] PASSED [  0%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-psycopg2] PASSED [  1%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-pg8000] PASSED [  1%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-psycopg2cffi] PASSED [  1%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-pypostgresql] PASSED [  2%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-pygresql] PASSED [  2%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-zxjdbc] PASSED [  2%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-mysqldb] PASSED [  3%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-pymysql] PASSED [  3%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-mysqlconnector] PASSED [  4%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-cymysql] PASSED [  4%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-oursql] PASSED [  4%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-gaerdbms] PASSED [  5%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-pyodbc] PASSED [  5%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-zxjdbc] PASSED [  5%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mssql-pyodbc] PASSED [  6%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mssql-mxodbc] PASSED [  6%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mssql-pymssql] PASSED [  6%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mssql-zxjdbc] PASSED [  7%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mssql-adodbapi] PASSED [  7%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_unsupported_db_type[oracle://...] PASSED [  8%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_unsupported_db_type[oracle+cx_oracle://...] PASSED [  8%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_unsupported_db_type[snowflake://...] PASSED [  8%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_unsupported_db_type[://...] PASSED [  9%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_unsupported_db_type[abcdefg] PASSED [  9%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_multiple_drivers PASSED [  9%]
tests/store/tracking/test_sqlalchemy_store.py::test_default_experiment ERROR [ 10%]
```

MySQL 9.0 is likely the culprit:

https://dev.mysql.com/doc/relnotes/mysql/9.0/en/news-9-0-0.html

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by harupy** [comment]
```
+ [[ ***mysql:3306/mlflowdb?charset=utf8mb4 == mssql* ]]
+ exec pytest tests/store/tracking/test_sqlalchemy_store.py tests/store/model_registry/test_sqlalchemy_store.py tests/db
Environment variable MLFLOW_TRACKING_URI is set to '***mysql:3306/mlflowdb?charset=utf8mb4', which may interfere with tests.
============================= test session starts ==============================
platform linux -- Python 3.8.19, pytest-8.2.2, pluggy-1.5.0 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /mlflow/home
configfile: pytest.ini
plugins: cov-5.0.0
collecting ... collected 273 items
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[sqlite-pysqlite] PASSED [  0%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[sqlite-pysqlcipher] PASSED [  0%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-psycopg2] PASSED [  1%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-pg8000] PASSED [  1%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-psycopg2cffi] PASSED [  1%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-pypostgresql] PASSED [  2%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-pygresql] PASSED [  2%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[postgresql-zxjdbc] PASSED [  2%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-mysqldb] PASSED [  3%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-pymysql] PASSED [  3%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-mysqlconnector] PASSED [  4%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-cymysql] PASSED [  4%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-oursql] PASSED [  4%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-gaerdbms] PASSED [  5%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-pyodbc] PASSED [  5%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mysql-zxjdbc] PASSED [  5%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mssql-pyodbc] PASSED [  6%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mssql-mxodbc] PASSED [  6%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mssql-pymssql] PASSED [  6%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mssql-zxjdbc] PASSED [  7%]
tests/store/tracking/test_sqlalchemy_store.py::test_correct_db_type_from_uri[mssql-adodbapi] PASSED [  7%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_unsupported_db_type[oracle://...] PASSED [  8%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_unsupported_db_type[oracle+cx_oracle://...] PASSED [  8%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_unsupported_db_type[snowflake://...] PASSED [  8%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_unsupported_db_type[://...] PASSED [  9%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_unsupported_db_type[abcdefg] PASSED [  9%]
tests/store/tracking/test_sqlalchemy_store.py::test_fail_on_multiple_drivers PASSED [  9%]
2024/07/03 10:37:40 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2002, "Can't connect to server on 'mysql' (115)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 0.1 seconds
2024/07/03 10:37:40 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2002, "Can't connect to server on 'mysql' (115)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 0.3 seconds
2024/07/03 10:37:40 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2002, "Can't connect to server on 'mysql' (115)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 0.7 seconds
2024/07/03 10:37:42 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2013, "Lost connection to server at 'handshake: reading initial communication packet', system error: 104")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 1.5 seconds
2024/07/03 10:37:43 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2002, "Can't connect to server on 'mysql' (115)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 3.1 seconds
2024/07/03 10:37:46 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2002, "Can't connect to server on 'mysql' (115)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 6.3 seconds
2024/07/03 10:37:53 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2013, "Lost connection to server at 'handshake: reading initial communication packet', system error: 104")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 12.7 seconds
2024/07/03 10:38:06 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2002, "Can't connect to server on 'mysql' (115)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 25.5 seconds
2024/07/03 10:39:05 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2005, "Unknown server host 'mysql' (-3)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 51.1 seconds
2024/07/03 10:40:30 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2005, "Unknown server host 'mysql' (-3)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in [102](https://github.com/mlflow/mlflow/actions/runs/9776298100/job/26988524501?pr=12560#step:5:103).3 seconds
2024/07/03 10:42:46 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2005, "Unknown server host 'mysql' (-3)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 204.7 seconds
2024/07/03 10:46:44 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(MySQLdb.OperationalError) (2005, "Unknown server host 'mysql' (-3)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Operation will be retried in 409.5 seconds
```

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -107,14 +107,15 @@ jobs:
...
+          docker images --filter reference=postgresql --filter reference=mysql --filter reference=mssql
           ./tests/db/compose.sh build --build-arg DEPENDENCIES="$(cat requirements/skinny-requirements.txt requirements/core-requirements.txt | grep -Ev '^(#|$)')"
       - name: Run tests
         run: |
           set +e
           err=0
           trap 'err=1' ERR
 
-          for service in $(./tests/db/compose.sh config --services | grep '^mlflow-')
+          for service in $(./tests/db/compose.sh config --services | grep '^mlflow-' | sort)
```
Comment (harupy) [inline .github/workflows/master.yml:118]: `sort` for a deterministic order

File: .github/workflows/master.yml [hunk 2]
```diff
@@ -107,14 +107,15 @@ jobs:
       - name: Build
         run: |
           ./tests/db/compose.sh pull -q postgresql mysql mssql
+          docker images
```
Comment (harupy) [inline .github/workflows/master.yml:110]: for easier debugging

File: tests/db/compose.yml [hunk 3]
```diff
@@ -36,7 +36,9 @@ services:
     command: tests/db/check_migration.sh
 
   mysql:
-    image: mysql
+    # TODO: MySQL 9.0 causes a hang. Investigate and re-enable.
```
Comment (harupy) [inline tests/db/compose.yml:39]:
```suggestion
    # TODO: MySQL 9.0 causes a hang. Investigate and remove pin.
```

File: mlflow/store/db/utils.py [hunk 4]
```diff
@@ -55,7 +55,7 @@
 
 _logger = logging.getLogger(__name__)
 
-MAX_RETRY_COUNT = 15
+MAX_RETRY_COUNT = 10
```
Comment (harupy) [inline mlflow/store/db/utils.py:58]: `MAX_RETRY_COUNT=15` results in sleeping for almost an hour in total, which is too long.

File: .github/workflows/master.yml (hunk 1 above)
Comment (xq-yin) [reply .github/workflows/master.yml:118]: oh good catch

**Review by xq-yin — APPROVED** [review]
LGTM

**Pull request merged**